samples/sobel/sobel_main.py

```diff
@@ -1,56 +1,231 @@
 import math
-import imageio
-from urllib.request import urlopen
 
+path = './images/harry.jpg'
 hcl.init(init_dtype=hcl.Float())
-img = Image.open(urlopen('http://i.stack.imgur.com/8zINU.gif'))
+img = Image.open(path)
 width, height = img.size
+Gx = hcl.placeholder((3, 3), "Gx")
+Gy = hcl.placeholder((3, 3), "Gy")
+A = hcl.placeholder((height, width, 3))
 
-A = hcl.placeholder((height,width), "A", dtype=hcl.Float())
-Gx = hcl.placeholder((3,3), "Gx",dtype=hcl.Float())
-Gy = hcl.placeholder((3,3), "Gy",dtype=hcl.Float())
+##############################################################################
+# Main Algorithm 
+# ==============
+# We perform a valid convolution with Sobel kernels. To perform convolution, 
+# we define the reduction axis on both convolution tensors to be a 3x3 region.
+# Then, we use the `.compute()` primitive that allows us to compute a new
+# tensor based on the function we give it.
 
 def sobel(A, Gx, Gy):
+    B = hcl.compute((height, width), lambda x, y: A[x][y][0] + A[x][y][1] + A[x][y][2], "B")
 
-   r = hcl.reduce_axis(0,3)
-   c = hcl.reduce_axis(0,3)
-   B = hcl.compute((height-2,width-2), 
-           lambda x,y: hcl.sum(A[x+r,y+c]*Gx[r,c], axis=[r,c]),
-           name="B", dtype=hcl.Float())
-   t = hcl.reduce_axis(0,3)
-   g = hcl.reduce_axis(0,3)
+    r = hcl.reduce_axis(0, 3)
+    c = hcl.reduce_axis(0, 3)
+    D = hcl.compute((height-2, width-2), 
+        lambda x, y: hcl.sum(B[x+r, y+c]*Gx[r, c], axis=[r, c], name = "sum1"), "D")
 
-   C = hcl.compute((height-2,width-2), 
-           lambda x,y: hcl.sum(A[x+t,y+g]*Gy[t,g], axis=[t,g]),
-           name="C", dtype=hcl.Float())
-   return hcl.compute((height-2,width-2), 
-              lambda x, y :hcl.sqrt(B[x,y]*B[x,y] + C[x,y]*C[x,y])/4328*255,
-              dtype=hcl.Float())
+    t = hcl.reduce_axis(0, 3)
+    g = hcl.reduce_axis(0, 3)
+    E = hcl.compute((height-2, width-2), 
+        lambda x, y: hcl.sum(B[x+t, y+g]*Gy[t, g], axis=[t, g], name = "sum2"), "E")
 
-s = hcl.create_schedule([A,Gx,Gy],sobel)
-f = hcl.build(s)
+    # constant factor to normalize the output
+    return hcl.compute((height-2,width-2), 
+        lambda x, y:hcl.sqrt(D[x][y]*D[x][y]+E[x][y]*E[x][y])*0.05891867, "Fimg")
 
-npA = np.array(img)
-npGx = np.array([[1,0,-1],[2,0,-2],[1,0,-1]])
-npGy = np.array([[1,2,1],[0,0,0],[-1,-2,-1]])
-hcl_A = hcl.asarray(npA)
+# create a schedule
+s = hcl.create_schedule([A, Gx, Gy], sobel)
+
+###############################################################################
+# Optimization
+# ============
+# HeteroCL provides different primitives to optimize the performance of the
+# program. In this example, since convolutions involve overlapped access to
+# the same memory location, we can make use of the idea of window- and line-
+# buffer to optimize for memory usage. The `reuse_at` primitive supports the
+# data reuse in the specified axis.
+# For further optimization, we apply the `partition` and `pipeline` primitives 
```

Comment:
Line- and window- buffers have to be used along with array partitioning to achieved expected throughput. Otherwise the memory contention will lead to II > 1. It is not "further optimization". Maybe it is better to rephrase a bit here.